gdal/frmts/netcdf/netcdfvirtual.h

```diff
@@ -0,0 +1,359 @@
+
+    /* netCDFVDimension
+     * -
+     * Contains the real dim id, real dimension name, and dimension length
+     */
+    class netCDFVDimension
+    {
+        friend class netCDFVID;
+
+        std::string real_dim_name;
+        int r_did = INVALID_DIM_ID;
+        int v_did;
+        size_t dim_len;
+        bool valid = true;
+
+        protected:
+            void setRealID(int realID) { this->r_did = realID; }
+            void invalidate();
+            void setLen(size_t len) { this->dim_len = len; }
+        public:
+                    netCDFVDimension(const char * name, size_t len, int dimid) :
+                        real_dim_name(name),
+                        v_did(dimid),
+                        dim_len(len)
+                    {}
+
+            std::string& getName() { return this->real_dim_name; }
+            size_t getLen() { return this->dim_len; }
+            int getRealID() { return this->r_did; }
+            int getVirtualID() { return this->v_did; }
+            bool isValid() { return this->valid; }
+    };
+
+    /* netCDFVVariable
+     * -
+     * Contains the variable name, variable type, etc.
+     */
+    class netCDFVVariable
+    {
+        friend class netCDFVID;
+
+        std::string real_var_name;
+        nc_type ntype;
+                int r_vid = INVALID_VAR_ID;
+        int v_vid;
+        int ndimc;
+        std::unique_ptr<int, std::default_delete<int[]>> dimid;
+        std::vector<std::shared_ptr<netCDFVAttribute>> attribs;
+        bool valid = true;
+
+        protected:
+            std::vector<std::shared_ptr<netCDFVAttribute>>& getAttributes() { return attribs; }
+            void invalidate();
+            void setRealID(int realID) { this->r_vid = realID; }
+        public:
+            netCDFVVariable(const char * name, nc_type xtype, int ndims, const int* dimidsp, int varid);
+            std::string& getName() { return real_var_name; }
+            int getRealID() { return r_vid; }
+            nc_type getType() { return ntype; }
+            int getDimCount() { return this->ndimc; }
+            const int* getDimIds() { return this->dimid.get(); }
+            bool isValid() { return this->valid; }
+    };
+
+    /* netCDFVID
+     * -
+     * A netCDF ID that sits on top of an actual netCDF ID
+     * And manages actual interaction with the real netCDF file
+     * A bit difference is that netCDFVDataset
+     * doesn't have fixed dim sizes, until defines are committed
+     *
+     * Also, virtual attributes only exist until the variable is committed. Use "real" attributes and "real" IDs
+     * for a variable after its been commited.
+     *
+     * ** Do not mix netCDF virtual dim and variable IDs with regular netCDF dim (a.k.a. "real")
+     * ids and variable ids. They are NOT necessarily compatible, and must be translated first, to be
+     * used in this manner **
+     */
+    class netCDFVID
+        {
+        int & ncid;
```

Comment:
does that needs to be a reference ?

Reply:
True, only dim growers need this. Since doesn't use dim growing maybe it doesn't have to be.
I guess I'm just kind of worried about the netCDF library sneakily deciding to change the netCDF ID which happened last few times. But sure we can try it without it.

Reply:
Hmm... it segfaults if I change it from reference.

I'll go investigate with GDB later